pkg/controller/scyllacluster/resource.go

```diff
@@ -895,3 +901,122 @@ func MakeRoleBinding(sc *scyllav1.ScyllaCluster) *rbacv1.RoleBinding {
 		},
 	}
 }
+
+func MakeJobs(sc *scyllav1.ScyllaCluster, services map[string]*corev1.Service, image string) ([]*batchv1.Job, error) {
+	var jobs []*batchv1.Job
+
+	cleanupJobs, err := makeCleanupJobs(sc, services, image)
+	if err != nil {
+		return nil, fmt.Errorf("can't make cleanup jobs: %w", err)
+	}
+	jobs = append(jobs, cleanupJobs...)
+
+	return jobs, nil
+}
+
+func makeCleanupJobs(sc *scyllav1.ScyllaCluster, services map[string]*corev1.Service, image string) ([]*batchv1.Job, error) {
+	var jobs []*batchv1.Job
+
+	for _, rack := range sc.Spec.Datacenter.Racks {
+		for i := int32(0); i < rack.Members; i++ {
+			svcName := naming.MemberServiceName(rack, sc, int(i))
+			svc, ok := services[svcName]
+			if !ok {
+				return nil, fmt.Errorf("service %q is missing", naming.ManualRef(sc.Namespace, svcName))
+			}
+
+			currentTokenRingHash, ok := svc.Annotations[naming.CurrentTokenRingHashAnnotation]
+			if !ok {
+				klog.V(4).Infof("Service %q is missing current token ring hash annotation", naming.ObjRef(svc))
+				continue
+			}
+
+			if len(currentTokenRingHash) == 0 {
+				klog.Warningf("Can't create cleanup Job for Service %s because it has unexpected empty current token ring hash annotation", klog.KObj(svc))
+				continue
+			}
+
+			latestTokenRingHash, ok := svc.Annotations[naming.LatestTokenRingHashAnnotation]
+			if !ok {
+				klog.V(4).Infof("Service %q is missing latest token ring hash annotation", naming.ObjRef(svc))
+				continue
+			}
+
+			if len(latestTokenRingHash) == 0 {
+				klog.Warningf("Can't create cleanup Job for Service %s because it has unexpected empty latest token ring hash annotation", klog.KObj(svc))
+				continue
+			}
+
+			if currentTokenRingHash == latestTokenRingHash {
+				klog.V(4).Infof("Node %q already cleaned up", naming.ObjRef(svc))
+				continue
+			}
+
+			klog.InfoS("Node requires a cleanup", "Node", naming.ObjRef(svc))
+
+			jobLabels := map[string]string{
+				naming.ClusterNameLabel: sc.Name,
+				naming.NodeJobLabel:     svcName,
+				naming.NodeJobTypeLabel: naming.JobTypeCleanup,
+			}
+			annotations := map[string]string{
+				naming.CleanupJobTokenRingHashAnnotation: currentTokenRingHash,
+			}
+
+			jobs = append(jobs, &batchv1.Job{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      naming.CleanupJobForService(svc),
+					Namespace: sc.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						*metav1.NewControllerRef(sc, scyllaClusterControllerGVK),
+					},
+					Labels:      jobLabels,
+					Annotations: annotations,
+				},
+				Spec: batchv1.JobSpec{
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{
+							Labels:      jobLabels,
+							Annotations: annotations,
+						},
+						Spec: corev1.PodSpec{
+							RestartPolicy: corev1.RestartPolicyOnFailure,
```

Comment:
what happens if the IP changes?

Reply:
cleanup job receives DNS name of member service